src/position.rs

```
@@ -0,0 +1,96 @@
+use crate::*;
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Position {
+    position_id: u128,
+    sell_token: AccountId,
+    buy_token: AccountId,
+    collateral_amount: Balance,
+    buy_token_price: Balance,
+    sell_token_price: Balance,
+    leverage: u128,
+}
+
+impl Position {
+    pub fn new(
+        position_id: u128,
+        sell_token: AccountId,
+        buy_token: AccountId,
+        collateral_amount: Balance,
+        buy_token_price: Balance,
+        sell_token_price: Balance,
+        leverage: u128,
+    ) -> Position {
+        Position {
+            position_id,
+            sell_token,
+            buy_token,
+            collateral_amount,
+            buy_token_price,
+            sell_token_price,
+            leverage,
+        }
+    }
+}
+
+#[near_bindgen]
+impl Contract {
+    #[private]
+    pub fn get_position_by_id(&self, position_id: U128) -> Position {
+        self.positions
+            .get(&position_id.0)
+            .unwrap_or_else(|| panic!("Position with current position_id: {}", position_id.0))
+    }
+
+    pub fn insert_position(&mut self, position: Position) {
+        self.positions.insert(&position.position_id, &position);
+
+        self.total_positions += 1;
+    }
+
+    pub fn open_position(
+        &mut self,
+        sell_token_amount: U128,
+        buy_token: AccountId,
+        sell_token: AccountId,
+        leverage: U128,
+    ) -> PromiseOrValue<U128> {
+        if self.user_profiles.get(&env::signer_account_id()).is_none() {
```

@dmytro-shash please always prefer the fail fast principle (aka precondition)